### DISCUSSION TITLE
Can we define the list of tracer names for model construction programatically?

### DISCUSSION THREAD

**Discussion by francispoulin** [CliMA/Oceananigans.jl#3650] [Experimental features]
I want to do a simulation with a lot of tracers, maybe 100 or more.  Rather than inputing each individually, I would really like to use a tracer that is a vector.  Is this possible?

If not, how much work would it be to get this working?

Thanks in advance!

**Comment by glwagner** [comment]
What do you mean by "inputting"? For post-processing, you can make a vector pretty easily:

```julia
vector_tracers = [c for c in model.tracers]
```

That just converts `model.tracers` from a `NamedTuple` to a `Vector`. You can output this by using a function for output,

```julia
function get_vector_tracers(sim)
    tracers = sim.model.tracers
    return [c for c in tracers]
end
```

If you want the tracers to be in a 4D array or something like that, and not just for post-processing but also while the simulation is running, that's also possible but a bit more work.

**Comment by francispoulin** [comment]
Thanks @glwagner for the reply and I am sorry I was not clear.

When we define the model, we must specify what tracers we want to use. For example, we could have `tracers = (:b, :P)` if we wanted to have buoyancy, and a tracer named P, say for phytoplankton.  This is easy but want if I wanted to have 100 different types of phytoplankton.  I could list `P1`, `P2`, ... 'P100`, but that would be very tedious. 

Instead, can we define a tracer to be a vector field and evolve each of those fields?  Maybe this is what you were thinking when you said a 4D array?

**Reply by glwagner** [reply]
Ah ok. You want to define the list of tracer _names_ programmatically (rather than writing out the names manually).

**Comment by jagoosw** [comment]
You could do `..., tracers = tuple([Symbol(:P, n) for n in 1:100]...), ...`

**Reply by francispoulin** [reply]
Thanks!  That works beautifully.  I will include a sample below in case others are interested in seeing what works.

```
grid = RectilinearGrid(size=128, z=(-0.5, 0.5), topology=(Flat, Flat, Bounded))
model = NonhydrostaticModel(; grid, tracers=([Symbol(:P, n) for n in 1:10]))
model
NonhydrostaticModel{CPU, RectilinearGrid}(time = 0 seconds, iteration = 0)
├── grid: 1×1×128 RectilinearGrid{Float64, Flat, Flat, Bounded} on CPU with 0×0×3 halo
├── timestepper: QuasiAdamsBashforth2TimeStepper
├── advection scheme: Centered reconstruction order 2
├── tracers: (P1, P2, P3, P4, P5, P6, P7, P8, P9, P10)
├── closure: Nothing
├── buoyancy: Nothing
└── coriolis: Nothing
```

**Reply by glwagner** [reply]
`tracers = [Symbol(:P, n) for n = 1:100]` will also work. Vectors of tracer names are supported for this purpose and also for iteratively building up the list of tracer names in a reasonably complicated model.

**Reply by glwagner** [reply]
Note that `tracers=([Symbol(:P, n) for n in 1:10])` and `tracers=[Symbol(:P, n) for n in 1:10]` are identical (in the former case, the surrounding parentheses are discarded during parsing).

**Reply by francispoulin** [reply]
Thanks!
